### DISCUSSION TITLE
Difference between `create_category_combinations` and `production.category_ids`

### DISCUSSION THREAD

**Discussion by gsaha009** [columnflow/columnflow#562] [Q&A]
Hello CF experts (@pkausw @mafrahm and others),

First of all, thank you guys for making the categorization smooth. I have one confusion though. If I have understood well, this `create_category_combinations` can create any possible categories of any depths (if > 1), and it also assigns corresponding IDs. So, all the new automatically produced categories should be stored somewhere, which are being used for Plotting I suppose. However, when I am calling `production.category_ids` at the end of my `selection` task, I can see that the intermediate categories are not assigned. 
e.g. following this structure:
```py
categories = {
        "channel": [config.get_category("tautau")],
        "abcd"   : [
            config.get_category("hadA"),  config.get_category("hadB"),
            config.get_category("hadA0"), config.get_category("hadB0"),
            config.get_category("hadC0"), config.get_category("hadD0"),
            config.get_category("hadC"),  config.get_category("hadD"),
        ],
        "TorF"   : [config.get_category("real_1"), config.get_category("fake_1")]
}
```
all possible categories are there while calling `create_category_combinations`, but with `production.category_ids`, only the categories combining root to leaf node are present, e.g. `tautau__hadA__real_1`, not like `tautau__hadA`.

Does the workflow follow this strategy? or I am doing something wrong.
any comment would be very helpful

** One more point: 
Maybe not related to it: 
If I use the entire root->leaf categories for plotting, it looks good, but not for any intermediate categories. 
[This](https://gsahacmspoint.web.cern.ch/CPinHToTauTau/Plots/2022/PreFit/Run3_2022PreEE_full_20241014_153211/plot__proc_8_a97f63528c__cat_mutau__lepD__var_hcand_1_eta.png) plot is for category `mutau__lepD`, and this category was produced using `create_category_combinations` method using a dictionary with these two types of sub-categories only. But if I follow the category hierarchy as mentioned above, for this same intermediate category the plot now looks like [this](https://gsahacmspoint.web.cern.ch/CPinHToTauTau/Plots/2022/PreFit/Run3_2022PreEE_full_20241027_020905/plot__proc_8_a421ab16f0__cat_mutau__lepD__var_hcand_1_eta.png). The data seems just double here. I am not certain if this observation is related to the above one or not. 

Regards,
Gourab

**Comment by pkausw** [comment]
Hi @gsaha009 ,

sorry for the late reply. Regarding the first question: You will never see intermediate categories with the `category_ids` producer because they are built at run time. That means that when you request a given intermediate (or rather: larger) category at the command line level, it's resolved into the smallest leafs first. This enables us to be very flexible with category definitions without rerunning the whole workflow every time - as long as there is no new leaf category, you won't need to run anything new. For more information, see for example the `How does this actually work?` box in the [category docs](https://columnflow.readthedocs.io/en/latest/user_guide/building_blocks/categories.html#create-a-category).

This approach has its dangers of course - for example, you can double-count events if the leaf categories aren't orthogonal. I'm not sure what happened with your plots you shared, but are you sure the selections of your `Categorizer` objects are 100% orthogonal? Can you share the code, or even better a minimial working example?

Cheers,
Philip

**Comment by gsaha009** [comment]
Hi @pkausw 

Thanks for the clarification
I must admit that I haven't got the 1st answer completely. But I have to think about it a bit
Regarding the second, all my categories are not orthogonal ofc. If you look at this [line](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/config/categories.py#L157), I have set these `real_1` and `fake_1` the same for data, so that I can get the histograms with data and `fake MC`, as I get from data and `prompt MC`. But, except these two categories, all other categories are orthogonal to each other.

Gourab

**Comment by pkausw** [comment]
Hi,

ok, then that's the problem - as stated in the docs, the leaf categories you use to build your larger categories must be orthogonal. Otherwise, you will double count events for sure. So a redefinition of the categories and there selections might help here.

Cheers,
Philip

**Reply by gsaha009** [reply]
Hi @pkausw 

I have a quick followup question regarding the same
Say in my category structure, I have defined one block like [this](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/config/categories.py#L81) and the other one is [this](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/config/categories.py#L121). Now, in the latest version, I am using `skip_existing = False` for the former, and `True` for the later one
All of the leaf categories must be orthogonal to each other as you mentioned. But, I want to plot `True MC` and `Fake MC` separately with the same `DATA`, Now, before jumping into the nested category structure, if I define two mother categories `tautau_TrueMC` and `tautau_FakeMC` which will be orthogonal to each other for MC, but the same for Data. Later, would it be right if I move ahead with the nested structure starting with these `tautau_TrueMC` and `tautau_FakeMC` categories?
If I am not wrong, I noticed that the double data issue was for intermediate categories, but those look good for categories created from root to leaf. Let me know you thoughts.
Thanks.

Gourab

**Reply by pkausw** [reply]
Hi @gsaha009 ,

I'm not sure if I understand what you actually want to do. But if you have defined two orthogonal regions `tautau_TrueMC` and `tautau_FakeMC` and are plotting separately (i.e. you don't plot a category `tautau` that would be the sum of the two) you should be good. You should be able to test this explicitely by e.g. looking at the yields of your `DATA` process for the different categories.

If you want to compare the two in one plot, you'll need to write some external makro for the plotting - we do not support plotting two different categories in the same plot. There are multiple ways to go about this of course, but you could in principle use these [tasks](https://github.com/columnflow/columnflow/blob/master/columnflow/tasks/inspection.py) that @mafrahm introduced some time ago to load the histograms as `Hist` objects and then do some custom plotting in the `run` function of the task.

Let me know if this helps!

**Reply by mafrahm** [reply]
Hi @gsaha009, I am not sure what your categorization is doing, but might it be the case that you implement some categories that could be better represented via processes?
It obviously depends on the usecase, but you could create process id's based on the existence of a generator-level tau and then create sub-processes corresponding to these taus.
There is no documentation for this, yet, but I for example implemented a separation between DY with and without heavy-flavoured jets in my analysis:
https://github.com/uhh-cms/hh2bbww/blob/e90add55c9c9b1b888b77a99d062349695c2b40d/hbw/production/process_ids.py#L202-L255
https://github.com/uhh-cms/hh2bbww/blob/e90add55c9c9b1b888b77a99d062349695c2b40d/hbw/config/processes.py#L168-L187

This mechanism also seamlessly integrates into our existing plotting tasks/functions, allowing you to plot your fake and real contributions into the same plot.
But if you need to separate all of your processes into fake and real tau's, this might be much more difficult/tedious to do.

Best
Mathis

**Reply by gsaha009** [reply]
Hi @pkausw and @mafrahm 

Thank you for the explanations. I am generating several processes from one process : From DY I am creating processes `Z to mm`, `Z to ee` etc. But my query here is a bit different 

We are estimating the jet to tau FR for H to mutau channel. Now, I need two sets of histograms. 
1. Data with all real/prompt MC for all possible categories
2. Data with all non-prompt/fake MC for the same

We will compute the ratio of several processes e.g. DY, W etc from the histograms mentioned in the second point, but the FR from option 1. But, I think that if I define this real and fake categories as processes for each process I am using, it is possible. Then I would have to run it twice maybe, for the plotting task only.

[Here](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/config/categories.py#L158-L179) I have added one type of category structure for tautau channel, where the root category is defined [here](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/categorization/main.py#L208). I also defined the ones for fake afterwards. There is only one catch: data is the same for both real and fake categories. But if I use the one for fake like the [real](https://github.com/gsaha009/CPinHToTauTau/blob/main/httcp/config/categories.py#L158-L179), would it be okay? 
before I was adding bith real and fake inside the big dictionary, so the data was summed up.

Gourab

**Comment by gsaha009** [comment]
Hi @pkausw 

I have restructured my categories. Although there are quite some duplicate categories now but I think the default option is to skip existing categories.
Thanks a lot :)

Gourab

**Comment by pkausw** [comment]
Hi @gsaha009 , since this isn't a bug but a feature, I transfered this issue to the discussions. If you are satisfied with the answer, feel free to mark the answer and close this discussion. If not, let us know!
